### ISSUE TITLE
Add support for zstd

### ISSUE THREAD

**Issue by essen** [essen/inet_tcp_compress_dist#2]
The best ratio/performance compression algorithm seems to be zstd (where lz4 is the best at performance when not caring too much about the ratio). It makes sense to add support for it as well and see how it fares.

Not much to see in the Erlang world around zstd: https://hex.pm/packages?_utf8=✓&search=zstd&sort=recent_downloads

The API we want is the streaming compression API: https://github.com/facebook/zstd/blob/dev/lib/zstd.h#L563

The Erlang NIF only supports the simple API.

The other driver has implemented the streaming compression API: https://github.com/chenzhuoyu/elixir-zstd/blob/master/c_src/nif_ex_zstd.c but the implementation does not quite fit:

* We want to accept `iovec()` as input for compression.
* As a result we want to produce `iovec()` when compressing, because we may end up yielding if we can't process everything at once.
* We also want to produce an `iovec()` when decompressing in order to avoid allocating too much; the driver does some weird sub-binary stuff instead.

The implementation overall would be very similar to the lz4 NIF so perhaps we should just adapt the lz4 NIF to zstd. We can always send a patch back to the Erlang NIF later on but the activity looks very low.

**Comment by gerhard** [comment]
This sounds like a good third option worth exploring. With this we will have tried deflate, lz4 & zstd. Pretty sure that whichever we end up choosing, it will be a solid choice. 👍

**Comment by essen** [comment]
I don't think it's a matter of us choosing anything, but more of providing different options to potential users. zstd would in this case be a much more efficient alternative to deflate and provide good compression at the expense of resources, while lz4 provides some compression with a much smaller footprint.

**Comment by essen** [comment]
Taken a second look to make sure it isn't a waste of time to create our own NIF. All existing points still stand, and I found two more.

The Elixir driver does not support Windows as well, we want to support it.

It also uses dirty CPU schedulers, that should not be necessary for our use case because we yield automatically when processing the iovec already. The dirty CPU scheduler can add latency and we don't want that by default, we only accept it when there is too much to process.

**Comment by essen** [comment]
https://github.com/rabbitmq/zstd-erlang

Need to add one more test for trying to stream compress/decompress multiple frames, which might not be necessary for this purpose, so better start by adding the Zstandard support to the compress dist first.

**Comment by essen** [comment]
I don't think the Elixir NIF is working right tbh, https://github.com/chenzhuoyu/elixir-zstd/blob/master/c_src/nif_ex_zstd.c#L248 would end up discarding data if the output buffer is too small. I've ran into the same issue when writing our NIF and that's now the one remaining open question, what size should we allocate? It's easy for LZ4 because it gives you the function, but no such thing here. I suppose if we compressed to an iovec it would not be a problem. I'll think about it next time.
